docs(apisix): wrote how-to-guide/../hashicorp-vault.md (#84)
api7/docs: apisix_versioned_docs/version-3.2.0/how-to-guide/security/secrets-management/hashicorp-vault.md

---
title: HashiCorp Vault
slug: /how-to-guide/security/secrets-management/hashicorp-vault
---

[HashiCorp Vault](https://www.vaultproject.io/) is a product that enables a centralized platform for managing secrets and encryption across different environments and applications. It provides a unified secrets management for storing and accessing, such as API keys, passwords, certificates, and more.

APISIX supports storing sensitive data in the configuration file as secrets, such as `admin_key`, etcd `username`, `password`.

This guide will show you how to configure HashiCorp Vault as a secrets manager, then store the APISIX `admin_key` in Vault and reference the key in APISIX configuration file.

![Integration with Vault](https://static.apiseven.com/uploads/2023/03/21/58qhtdBI_Vault.jpg)

## Prerequisite(s)

* Install [Docker](https://docs.docker.com/get-docker/).
* Install [cURL](https://curl.se/) to send requests to the services for validation.

> * Install [ZIP](https://infozip.sourceforge.net/Zip.html) to uncompress the Vault binary from the [official distributed zipped file](https://developer.hashicorp.com/vault/downloads).

Reviewer: > uncompress
unzip

[//]: <TODO: Fix the insecure `allow_admin` in future>

```shell
docker exec apisix-quickstart /bin/sh -c "echo '
deployment:
  role: traditional
  role_traditional:
    config_provider: etcd
  admin:
    allow_admin:
      - 0.0.0.0/0
      - 127.0.0.0/24
    admin_key:
      -
        name: admin
        # highlight-start
        key: {{ with secret \"secret/apisix\"}} {{ .Data.data.adminKey }} {{ end }}
        # highlight-end
        role: admin
plugin_attr:
  prometheus:
    export_addr:
      ip: 0.0.0.0
      port: 9091
' > /usr/local/apisix/conf/config.ctmpl"
```

The following command starts the Vault Agent daemon in the APISIX instance with the configuration file `vault-agent-apisix.hcl` created above. The Agent automatically gets the secrets and injects them into the APISIX configuration file `config.yaml`.

```shell
docker exec -d apisix-quickstart vault agent -config=/usr/local/apisix/conf/vault-agent-apisix.hcl
```

## Validate

Since the `admin_key` is automatically injected into the default APISIX configuration file `/usr/local/apisix/conf/config.yaml`, you need to reload your APISIX instance for configuration changes to take effect:

```shell

> docker exec apisix-quickstart apisix reload

Reviewer: Should we mention this `reload` is actually a hot reload?

  }
}

> // authorization with token

Reviewer 2: ```suggestion
// authentication with token
```

' > /usr/local/apisix/conf/vault-agent-apisix.hcl"
```

> Vault Agent injects secrets into the APISIX configuration file `config.yaml` according to the rendering template file `config.ctmpl` after it's started. Both two files are stored in the APISIX default configuration path `/usr/local/apisix/conf/`.

Reviewer 2: ```suggestion
Vault Agent injects secrets into the APISIX configuration file `config.yaml` according to the rendered template file `config.ctmpl` after it's started. Both files are stored in the APISIX default configuration path `/usr/local/apisix/conf/`.
```

Author: Remove it may be better.

' > /usr/local/apisix/conf/config.ctmpl"
```

> The following command starts the Vault Agent daemon in the APISIX instance with the configuration file `vault-agent-apisix.hcl` created above. The Agent automatically will gets the secrets and injects them into the APISIX configuration file `config.yaml`.

Reviewer 2: ```suggestion
The following command starts the Vault Agent daemon in the APISIX instance with the configuration file `vault-agent-apisix.hcl` created above. The Agent automatically gets the secrets and injects them into the APISIX configuration file `config.yaml`.
```

docker exec apisix-quickstart apisix reload
```

> Verify that the new `admin_key` is in effect:

Reviewer 2: ```suggestion
Verify that the new `admin_key` is in effect:

1. Request Admin API with the correct `admin_key`:
```

{"list":[],"total":0}
```

> Try to request Admin API with the wrong `admin_key`:

Reviewer 2: ```suggestion
2. Request Admin API with a wrong `admin_key`:
```

```shell
curl -i "http://localhost:9180/apisix/admin/routes" -H 'X-API-KEY: wrong-key'
```

The expected response is similar to the following:

```
HTTP/1.1 401 Unauthorized
Date: Thu, 23 Mar 2023 08:37:04 GMT
Content-Type: application/json
Transfer-Encoding: chunked
Connection: keep-alive
Server: APISIX/3.2.0

> Access-Control-Allow-Origin: *

Reviewer 2: Some unnecessary information can be removed?

Access-Control-Allow-Credentials: true
Access-Control-Expose-Headers: *
Access-Control-Max-Age: 3600

{"error_msg":"failed to check token"}
```

## Next Steps

> APISIX supports Vault as the backend to store the sensitive data used in authorization and certification.

Reviewer 2: authorization or authentication?

> To learn more about integrate Vault with APISIX for authentication and certification, see other guides in the same directory as this guide. (coming soon).

Reviewer 2: ```suggestion
To learn more about integrating Vault with APISIX for authentication and certification, see other guides in the same directory as this guide. (coming soon).
```
